### DISCUSSION TITLE
How to use AutoPath as a type

### DISCUSSION THREAD

**Discussion by co-sic** [mikro-orm/mikro-orm#3027] [Q&A] [answered]
I want to use the AutoPath type pretty much the same as it is used in the Populate type.
Let's say i have an entity `Customer` i want to define a path like this:
```const path: AutoPath<Customer> = "customerNumber";```
But without the second generic it is not working. What do i need to pass as the second generic, if i want to be able to pass all of the properties of the fields of Customer and all possible subfields like `companyAddress.companyName`? I checked the source, but the type is very complex and hard to read for me.

**Comment by B4nan** [comment, marked as answer]
It should be the literal type of the string you are passing it, e.g. `AutoPath<Customer, 'customerNumber'>`. For your use case you might need something a bit different, as aim of `AutoPath` is not (directly) validation, but rather intellisense/autocomplete - it needs this type argument (whose value is always inferred) so it knows what to do - what to suggest (what nested relations are defined). I am not sure if its even possible what you want, but this type is rather community approach, it took me quite some time (weeks) to get it working properly (and there are still some edge cases that dont work 100% well), so I am no expert in this string literal magic :] 

You would probably need a type that would generate all possible paths (and a way to get around recursion). `AutoPath` validates what you gives it (in this second type arg) and offers extensions of it.

**Reply by co-sic** [reply]
Thank you for the explanation! One last question, as i'm not sure if i'm missing something here or not understanding this correctly. If use the entityManager to find something like this
```orm.entityManager.find(Customer, {}, { populate: ['companyAddress.companyName'] });```
Then the type of the populate array works exactly as i want it. It validates and it has autocompletion/itellisence. And i don't have to input any literal type here as second type argument. So how does this work, and can't i use this somehow for my usecase? If it's to complicated you can just tell me that its not working for my use case and i can accept that :D

**Reply by B4nan** [reply]
This works based on inferrence, the `companyAddress.companyName` is inferred as the second type argument. The `AutoPath` type gets this information based on the value you provide to `populate`, and offers both validation as well as intellisense to its extensions (nested relations). But for your use case you don't want to infer anything (or rather you can't infer anything this way - if you want to specify a variable type and assign it a value, there is nothing to infer, the type and the value are separate).

So if you are in a context of some method, use the type argument and it will work. But if you want what to type a variable as in the OP, I don't think this type can be used for that, definitely not directly. I believe you would need a different approach, that basically generates all possible properties (and solve the recursion problem somehow). But as said above, I don't feel like an expert in string literal types, maybe there is some easy way I am missing. Try asking on stack overflow, that is [how I initially got this working](https://stackoverflow.com/questions/68302695/ts-strict-paths-with-dot-notation-in-array), plus a lots of googling and trial/failure :]

**Reply by co-sic** [reply]
Thank your for the pointers, that helped me to understand it! In the end i went for the same approach you use with the entityManger find methods. I have a class for MikroOrmPagination (generic mapping of graphql query param to mikro-orm query params) and i pass this a definition of possible searchColumns where i wanted to use the AutoPath type instead of an untyped string. I had some other generics that i was using and if i learned with some try and error that you have to either supply all generics (which i could not do because of AutoPath) or supply non at all at let infer do it's magic. So it the end i went from
```
const pagination: MikroOrmPagination<Entity, ...> = new MikroOrmPagination<Entity, ...>([{path: "address.companyName", ...}]);
```
to
```
const pagination = new MikroOrmPagination(Entity, ..., [{path: "address.companyName", ...}]);
```
and now it is working like a charm, thanks!
